ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java

```diff
@@ -17,38 +17,83 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 
 /**
  * A HTTP/2 implementation of {@link NetworkClient}.
- * TODO: More implementation details will be added in next PR.
  */
 public class Http2NetworkClient implements NetworkClient {
   private static final Logger logger = LoggerFactory.getLogger(Http2NetworkClient.class);
   private final EventLoopGroup eventLoopGroup;
   private final ChannelPoolMap<InetSocketAddress, ChannelPool> pools;
+  private Http2ClientResponseHandler http2ClientResponseHandler;
+  public static AttributeKey<RequestInfo> REQUEST_INFO = AttributeKey.newInstance("RequestInfo");
 
   public Http2NetworkClient(Http2ClientMetrics http2ClientMetrics, Http2ClientConfig http2ClientConfig,
       SSLFactory sslFactory) {
     this.eventLoopGroup = new NioEventLoopGroup(http2ClientConfig.http2NettyEventLoopGroupThreads);
     this.pools = new Http2ChannelPoolMap(sslFactory, eventLoopGroup, http2ClientConfig);
+    this.http2ClientResponseHandler = new Http2ClientResponseHandler();
   }
 
   @Override
   public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop,
       int pollTimeoutMs) {
-    return null;
+    long startTime = SystemTime.getInstance().milliseconds();
+    List<ResponseInfo> readyResponseInfo = new ArrayList<>();
+    // Send request
+    for (RequestInfo requestInfo : requestsToSend) {
+      requestInfo.streamAcquireTime = SystemTime.getInstance().milliseconds();
+      this.pools.get(InetSocketAddress.createUnresolved(requestInfo.getHost(), requestInfo.getPort().getPort()))
+          .acquire()
+          .addListener((GenericFutureListener<Future<Channel>>) future -> {
+            if (future.isSuccess()) {
+              logger.trace(
+                  "Stream acquired: " + (SystemTime.getInstance().milliseconds() - requestInfo.streamAcquireTime));
+              Channel streamChannel = future.getNow();
+              streamChannel.pipeline().addLast(http2ClientResponseHandler);
+              streamChannel.pipeline().addLast(new AmbrySendToHttp2Adaptor());
+              streamChannel.attr(REQUEST_INFO).set(requestInfo);
+              streamChannel.write(requestInfo.getRequest());
+              requestInfo.streamSendTime = SystemTime.getInstance().milliseconds();
+            } else {
+              logger.error("Couldn't acquire stream channel: " + future.cause());
+              readyResponseInfo.add(new ResponseInfo(requestInfo, NetworkClientErrorCode.NetworkError, null));
+            }
+          });
+    }
+    logger.trace("HTTP2 send request time: " + (SystemTime.getInstance().milliseconds() - startTime));
+    // TODO: close stream channel for requestsToDrop. Need a hashmap from corelationId to streamChannel
+
+    // Add the good responses to readyResponseInfo
+    Queue<ResponseInfo> queue = http2ClientResponseHandler.getQueueToConsume();
+    ResponseInfo responseInfo = queue.poll();
+    while (responseInfo != null) {
```

Comment:
To use one queue, I also prefer `LinkedBlockingQueue ` because it has a method `drainTo`, which takes all current available elements. This fits our use case. 

(To make sure we are in the same page): The busy loop happens when there is no/low request. 
```
Router.run { 
     while() {
         ....
         client.sendAndPool()
         ....
     }
}
```
So either one queue or two queues, we need a condition wait in `sendAndPool()` and be waken up when responses come.

Reply:
Yep, I think we are on the same page. For a blocking queue example, you can look at `LocalRequestResponseChannel`. The only missing part is the implementation of `wakeup`, but I think that is doable with a method that puts a dummy element in the queue.

With the 2 queue system, you would need a condition variable to signal when the list is not empty.

Reply:
@cgtz , I am changing to use one blocking queue. But still, I think a cv wait needed when there is no response in h2 client and a signal should be sent when a h2 response arrive, right?

Reply:
I think there may be a way around having an additional condition variable, since LinkedBlockingQueue already has one inside. It already supports doing a timed wait and getting notified when a new element is present. The only case that would have to be handled differently would be the wakeup call. Here is an idea I had for how to make it work. Tell me if it makes sense (adapting some code from LocalRequestResponseChannel):

For the NetworkClient (consumer) side:
```java
  private static final ResponseInfo WAKEUP_MARKER = new ResponseInfo(null, null, null);
  BlockingQueue<ResponseInfo> responseQueue = new LinkedBlockingQueue<>();
  
  List<ResponseInfo> sendAndPoll() {
    // handle requests
    
    ResponseInfo firstResponse = null;
    try {
      firstResponse = responseQueue.poll(pollTimeoutMs, TimeUnit.MILLISECONDS);
    } catch (InterruptedException ie) {
      logger.debug("Interrupted polling responses");
    }
    if (firstResponse == null) {
      return Collections.emptyList();
    }
    // using a linked list to make removal easier (maybe there's a better way to handle this)
    List<ResponseInfo> responseList = new LinkedList<>();
    responseList.add(firstResponse);
    responseQueue.drainTo(responseList);
    // remove any wakeup markers, since they are just there to signal this thread to unblock itself
    responseList.removeIf(responseInfo -> responseInfo == WAKEUP_MARKER);
    return responseList;
  }

  void wakeup() {
    // if sendAndPoll is currently executing a timed poll on the blocking queue, we need to put something there
    // to wake it up before a response comes
    try {
      responseQueue.put(WAKEUP_MARKER);
    } catch (InterruptedException e) {
      logger.debug("Interrupted while waking up");
    }
  }
```

For the ResponseHandler (producer) side:
```java
  void produce(ResponseInfo responseInfo) throws InterruptedException {
    responseQueue.put(responseInfo);
  }
```

Reply:
Cool idea. Will use it.